File: backend/semantic_cache.py
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Optional

import httpx
from supabase import Client

from backend.usage import estimate_cost

EMBED_URL = "https://openrouter.ai/api/v1/embeddings"
EMBED_MODEL = "openai/text-embedding-3-small"
SIMILARITY_THRESHOLD = 0.92
TTL_DAYS = 30
TABLE = "semantic_cache"


def last_user_question(messages: List[Dict[str, Any]]) -> str:
    for item in reversed(messages or []):
        if item.get("role") == "user":
            return str(item.get("content") or "").strip()
    return ""


def normalize_question(question: str) -> str:
    return " ".join((question or "").lower().split())[:4000]


def embedding_to_vector(values: List[float]) -> str:
    return "[" + ",".join(f"{value:.8f}" for value in values) + "]"


async def embed_question(api_key: str, question: str) -> Optional[List[float]]:
    text = (question or "").strip()
    if not text or not api_key:
        return None
    try:
        async with httpx.AsyncClient(timeout=20.0) as client:
            response = await client.post(
                EMBED_URL,
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                    "HTTP-Referer": "http://localhost:8001",
                    "X-Title": "Proxy FinOps AI",
                },
                json={"model": EMBED_MODEL, "input": text[:8000]},
            )
        if response.status_code != 200:
            print(f"semantic_cache embed HTTP {response.status_code}: {response.text[:180]}")
            return None
        data = response.json()
        vector = (data.get("data") or [{}])[0].get("embedding")
        if isinstance(vector, list) and vector:
            return [float(item) for item in vector]
    except Exception as exc:
        print(f"semantic_cache embed failed: {exc}")
    return None


def lookup_exact(supabase: Optional[Client], workspace_id: Optional[str], question: str) -> Optional[Dict[str, Any]]:
    if not supabase or not workspace_id:
        return None
    norm = normalize_question(question)
    if not norm:
        return None
    try:
        result = (
            supabase.table(TABLE)
            .select("id, question, answer, requested_model, served_model, prompt_tokens, completion_tokens, hit_count")
            .eq("workspace_id", workspace_id)
            .eq("question_norm", norm)
            .gt("expires_at", datetime.now(timezone.utc).isoformat())
            .order("created_at", desc=True)
            .limit(1)
            .execute()
        )
        if result.data:
            return result.data[0]
    except Exception as exc:
        print(f"semantic_cache exact lookup skipped: {exc}")
    return None


def lookup_similar(
    supabase: Optional[Client],
    workspace_id: Optional[str],
    embedding: List[float],
) -> Optional[Dict[str, Any]]:
    if not supabase or not workspace_id or not embedding:
        return None
    try:
        result = supabase.rpc(
            "match_semantic_cache",
            {
                "p_workspace_id": workspace_id,
                "p_embedding": embedding_to_vector(embedding),
                "p_threshold": SIMILARITY_THRESHOLD,
                "p_limit": 1,
            },
        ).execute()
        if result.data:
            return result.data[0]
    except Exception as exc:
        print(f"semantic_cache similar lookup skipped: {exc}")
    return None


def mark_cache_hit(supabase: Optional[Client], row: Dict[str, Any]) -> None:
    if not supabase or not row.get("id"):
        return
    try:
        supabase.table(TABLE).update({"hit_count": int(row.get("hit_count") or 0) + 1}).eq("id", row["id"]).execute()
    except Exception as exc:
        print(f"semantic_cache hit_count skipped: {exc}")


def store_semantic_entry(
    supabase: Optional[Client],
    *,
    workspace_id: Optional[str],
    question: str,
    answer: str,
    embedding: Optional[List[float]],
    requested_model: str,
    served_model: str,
    prompt_tokens: int,
    completion_tokens: int,
) -> None:
    if not supabase or not workspace_id:
        return
    norm = normalize_question(question)
    if not norm or not (answer or "").strip():
        return
    expires = datetime.now(timezone.utc) + timedelta(days=TTL_DAYS)
    row: Dict[str, Any] = {
        "workspace_id": workspace_id,
        "question": question[:8000],
        "question_norm": norm,
        "answer": answer,
        "requested_model": requested_model,
        "served_model": served_model,
        "prompt_tokens": prompt_tokens,
        "completion_tokens": completion_tokens,
        "expires_at": expires.isoformat(),
    }
    if embedding:
        row["embedding"] = embedding_to_vector(embedding)
    try:
        supabase.rpc("cleanup_semantic_cache").execute()
    except Exception:
        pass
    try:
        supabase.table(TABLE).insert(row).execute()
    except Exception as exc:
        print(f"semantic_cache store skipped: {exc}")


def cache_hit_costs(requested_model: str, prompt_tokens: int, completion_tokens: int) -> Dict[str, float]:
    without = estimate_cost(requested_model, prompt_tokens, completion_tokens)
    return {
        "cost_with_proxy": 0.0,
        "cost_without_proxy": without,
        "savings_usd": without,
    }


async def resolve_semantic_lookup(
    supabase: Optional[Client],
    *,
    workspace_id: Optional[str],
    api_key: str,
    question: str,
) -> tuple[Optional[Dict[str, Any]], Optional[List[float]]]:
    if not question:
        return None, None
    hit = lookup_exact(supabase, workspace_id, question)
    if hit:
        mark_cache_hit(supabase, hit)
        return hit, None
    embedding = await embed_question(api_key, question)
    similar = lookup_similar(supabase, workspace_id, embedding or [])
    if similar:
        mark_cache_hit(supabase, similar)
        return similar, embedding
    return None, embedding

File: dashboard/src/lib/semantic-cache.ts
import { estimateCost, openrouterHeaders } from "@/lib/smart-balance";

const EMBED_URL = "https://openrouter.ai/api/v1/embeddings";
const EMBED_MODEL = "openai/text-embedding-3-small";
const SIMILARITY_THRESHOLD = 0.92;
const TTL_DAYS = 30;

export type SemanticHit = {
  id: string;
  question: string;
  answer: string;
  requested_model?: string | null;
  served_model?: string | null;
  prompt_tokens: number;
  completion_tokens: number;
  similarity?: number;
  embedding?: number[];
};

type AdminClient = {
  from: (table: string) => any;
  rpc: (name: string, args: Record<string, unknown>) => any;
};

export function lastUserQuestion(messages: { role?: string; content?: string }[]) {
  for (let index = messages.length - 1; index >= 0; index -= 1) {
    if (messages[index]?.role === "user") {
      return (messages[index].content || "").trim();
    }
  }
  return "";
}

export function normalizeQuestion(question: string) {
  return question.toLowerCase().replace(/\s+/g, " ").trim().slice(0, 4000);
}

function embeddingToVector(values: number[]) {
  return `[${values.map((value) => value.toFixed(8)).join(",")}]`;
}

export async function embedQuestion(apiKey: string, question: string): Promise<number[] | null> {
  const text = question.trim();
  if (!text || !apiKey) return null;
  try {
    const response = await fetch(EMBED_URL, {
      method: "POST",
      headers: openrouterHeaders(apiKey),
      body: JSON.stringify({ model: EMBED_MODEL, input: text.slice(0, 8000) }),
    });
    if (!response.ok) {
      console.error("semantic_cache embed HTTP", response.status, (await response.text()).slice(0, 180));
      return null;
    }
    const data = (await response.json()) as { data?: { embedding?: number[] }[] };
    const vector = data.data?.[0]?.embedding;
    return Array.isArray(vector) && vector.length ? vector : null;
  } catch (error) {
    console.error("semantic_cache embed failed", error);
    return null;
  }
}

export async function lookupExactCache(
  admin: AdminClient,
  workspaceId: string,
  question: string,
): Promise<SemanticHit | null> {
  const norm = normalizeQuestion(question);
  if (!norm) return null;
  const { data, error } = await admin
    .from("semantic_cache")
    .select("id, question, answer, requested_model, served_model, prompt_tokens, completion_tokens, hit_count")
    .eq("workspace_id", workspaceId)
    .eq("question_norm", norm)
    .gt("expires_at", new Date().toISOString())
    .order("created_at", { ascending: false })
    .limit(1)
    .maybeSingle();
  if (error || !data) return null;
  return data as SemanticHit;
}

export async function lookupSimilarCache(
  admin: AdminClient,
  workspaceId: string,
  embedding: number[],
): Promise<SemanticHit | null> {
  const { data, error } = await admin.rpc("match_semantic_cache", {
    p_workspace_id: workspaceId,
    p_embedding: embeddingToVector(embedding),
    p_threshold: SIMILARITY_THRESHOLD,
    p_limit: 1,
  });
  if (error || !data?.[0]) return null;
  return data[0] as SemanticHit;
}

export async function resolveSemanticLookup(
  admin: AdminClient,
  workspaceId: string,
  apiKey: string,
  question: string,
): Promise<{ hit: SemanticHit | null; embedding: number[] | null }> {
  if (!question) return { hit: null, embedding: null };
  const exact = await lookupExactCache(admin, workspaceId, question);
  if (exact) {
    await admin
      .from("semantic_cache")
      .update({ hit_count: Number((exact as SemanticHit & { hit_count?: number }).hit_count || 0) + 1 })
      .eq("id", exact.id);
    return { hit: exact, embedding: null };
  }
  const embedding = await embedQuestion(apiKey, question);
  if (!embedding) return { hit: null, embedding: null };
  const similar = await lookupSimilarCache(admin, workspaceId, embedding);
  if (similar) {
    await admin
      .from("semantic_cache")
      .update({ hit_count: Number((similar as SemanticHit & { hit_count?: number }).hit_count || 0) + 1 })
      .eq("id", similar.id);
    return { hit: similar, embedding };
  }
  return { hit: null, embedding };
}

export async function storeSemanticEntry(
  admin: AdminClient,
  options: {
    workspaceId: string;
    question: string;
    answer: string;
    embedding?: number[] | null;
    requestedModel: string;
    servedModel: string;
    promptTokens: number;
    completionTokens: number;
  },
) {
  const norm = normalizeQuestion(options.question);
  if (!norm || !options.answer.trim()) return;
  const expires = new Date(Date.now() + TTL_DAYS * 24 * 60 * 60 * 1000).toISOString();
  await admin.rpc("cleanup_semantic_cache", {});
  const { error } = await admin.from("semantic_cache").insert({
    workspace_id: options.workspaceId,
    question: options.question.slice(0, 8000),
    question_norm: norm,
    answer: options.answer,
    embedding: options.embedding ? embeddingToVector(options.embedding) : null,
    requested_model: options.requestedModel,
    served_model: options.servedModel,
    prompt_tokens: options.promptTokens,
    completion_tokens: options.completionTokens,
    expires_at: expires,
  });
  if (error) {
    console.error("semantic_cache store failed", error);
  }
}

export function cacheHitCosts(requestedModel: string, promptTokens: number, completionTokens: number) {
  const costWithoutProxy = estimateCost(requestedModel, promptTokens, completionTokens);
  return {
    costWithProxy: 0,
    costWithoutProxy,
    savingsUsd: costWithoutProxy,
  };
}
